src/dbnode/storage/bootstrap/bootstrapper/persist.go

```diff
@@ -0,0 +1,314 @@
+	"github.com/m3db/m3/src/m3ninx/doc"
+	"github.com/m3db/m3/src/m3ninx/index"
+	"github.com/m3db/m3/src/m3ninx/index/segment"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+// SharedPersistManager is a lockable persist manager that's safe to be shared across threads.
+type SharedPersistManager struct {
+	sync.Mutex
+	Mgr persist.Manager
+}
+
+// SharedCompactor is a lockable compactor that's safe to be shared across threads.
+type SharedCompactor struct {
+	sync.Mutex
+	Compactor *compaction.Compactor
+}
+
+// PersistBootstrapIndexSegment is a helper function that persists bootstrapped index segments for a ns -> block of time.
+func PersistBootstrapIndexSegment(
+	ns namespace.Metadata,
+	requestedRanges result.ShardTimeRanges,
+	indexResults result.IndexResults,
+	indexBuilders result.IndexBuilders,
+	persistManager *SharedPersistManager,
+	resultOpts result.Options,
+) error {
+	// If we're performing an index run with persistence enabled
+	// determine if we covered a full block exactly (which should
+	// occur since we always group readers by block size).
+	min, max := requestedRanges.MinMax()
+	blockSize := ns.Options().IndexOptions().BlockSize()
+	blockStart := min.Truncate(blockSize)
+	blockEnd := blockStart.Add(blockSize)
+	expectedRangeStart, expectedRangeEnd := blockStart, blockEnd
+
+	// Index blocks can be arbitrarily larger than data blocks, but the
+	// retention of the namespace is based on the size of the data blocks,
+	// not the index blocks. As a result, it's possible that the block start
+	// for the earliest index block is before the earliest possible retention
+	// time.
+	// If that is the case, then we snap the expected range start to the
+	// earliest retention block start because that is the point in time for
+	// which we'll actually have data available to construct a segment from.
+	//
+	// Example:
+	//  Index block size: 4 hours
+	//  Data block size: 2 hours
+	//  Retention: 6 hours
+	//           [12PM->2PM][2PM->4PM][4PM->6PM] (Data Blocks)
+	// [10AM     ->     2PM][2PM     ->     6PM] (Index Blocks)
+	retentionOpts := ns.Options().RetentionOptions()
+	nowFn := resultOpts.ClockOptions().NowFn()
+	earliestRetentionTime := retention.FlushTimeStart(retentionOpts, nowFn())
+	if blockStart.Before(earliestRetentionTime) {
+		expectedRangeStart = earliestRetentionTime
+	}
+
+	shards := make(map[uint32]struct{})
+	expectedRanges := make(result.ShardTimeRanges, len(requestedRanges))
+	for shard := range requestedRanges {
+		shards[shard] = struct{}{}
+		expectedRanges[shard] = xtime.Ranges{}.AddRange(xtime.Range{
+			Start: expectedRangeStart,
+			End:   expectedRangeEnd,
+		})
+	}
+
+	indexBlock, ok := indexResults[xtime.ToUnixNano(blockStart)]
+	if !ok {
+		return fmt.Errorf("did not find index block for blocksStart: %d", blockStart.Unix())
+	}
+	builder, ok := indexBuilders[xtime.ToUnixNano(blockStart)]
+	if !ok {
+		// No-op if there are no index builders for this time block (nothing to persist).
+		return nil
```

Comment:
Same here, wonder if we should error out in this case? I would expect if we call BuildBootstrapIndexSegment(...) we really do want to actually build the segment?

Maybe we should avoid this being a precondition.

Reply:
Looks like we still need to return an error here I think?